File: typescript/src/patterns/Builder.ts
export class Product {
    public name: String;
    public value: Number;
    public description: String;

    constructor (){
        this.name = "";
        this.value = 0;
        this.description = "";
    }
}

export class Builder {
    private product: Product;
    private products: Product[] = [];

    constructor () {
        this.product = new Product();
    }

    public reset() {
        this.product = new Product();
    }

    public setName(name: String) {
        this.product.name = name;
    }

    public setValue(value: Number) {
        this.product.value = value;
    }

    public setDescriptions(description: String) {
        this.product.description = description;
    }

    public build() {
        this.products.push(this.product);
        this.reset();
    }

    public getProducts(): Product[] {        
        return this.products;
    }
}

File: java/Builder.java
class Product {
    private String name;
    private float value;
    private String description;

    private Product(Builder builder) {
        this.name = builder.name;
        this.value = builder.value;
        this.description = builder.description;
    }

    public static class Builder {
        private String name;
        private float value;
        private String description;

        public Builder setName(String name) {
            this.name = name;
            return this;
        }
        
        public Builder setValue(float value) {
            this.value = value;
            return this;
        }
        
        public Builder setDescription(String description) {
            this.description = description;
            return this;
        }

        public Product build() {
            return new Product(this);
        }
    }

    public String toString() {
        return "Name: " + this.name +
               "\nDescriptions: " + this.description +
               "\nValue: " + this.value;

    }

    public static void main(String[] args) {        
        Product product = new Product.Builder()
        .setName("Bierland")
        .setDescription("Cerveja")
        .setValue(10)
        .build();

        System.out.println(product.toString());
    }
}